### PR TITLE
HHH-12907 Avoid creating too many ByteBuddy objects

### PR DESCRIPTION
**Pull request by gsmet**
This commit should reduce the garbage collection pressure as reported at
https://github.com/raphw/byte-buddy/issues/515.

 * https://hibernate.atlassian.net/browse/HHH-12907

### CODE REVIEW

**Comment by Sanne** [comment]
looks nice. Running tests...

**Comment by Sanne** [comment]
Passed tests ready to merge, but not merging myself as @gsmet has some more tuning in mind.

**Comment by gsmet** [comment]
@Sanne I force-pushed some changes:
- considered the additional filters you mentioned on HipChat
- centralized these elements in an object: it's more readable and it will be easier to nullify it if we implement some post-bootstrap "cache" cleanup in the future

File: hibernate-core/src/main/java/org/hibernate/bytecode/internal/bytebuddy/ByteBuddyState.java [hunk 1]
```diff
@@ -52,6 +65,8 @@
 	private final ForDeclaredMethods getDeclaredMethodMemberSubstitution;
 	private final ForDeclaredMethods getMethodMemberSubstitution;
 
+	private final ProxyDefinitionHelpers proxyDefinitionHelpers;
```
[the remark this answers is not in the document]
Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/bytecode/internal/bytebuddy/ByteBuddyState.java]: I thought the same. yet it looks like you'd be able to easily change that in future w/o changing any SPI as the accessor is not static.

**Pull request merged**